### PR TITLE
Move block store CRUD functionality from the node store component in casper-storage.

### PR DESCRIPTION
**Pull request by alsrdn**
Moved the basic CRUD block store operations to casper-storage. This allows other crates to interact with the node storage databases without the need to go through the storage component of the node.

The `block_store` module in casper-storage now provides traits to interact with a casper block store. It also provides concrete implementations for using LMDB as the backing store and also allows accessing indexed data.

Some types that were defined in the node have moved to casper-types and casper-storage as a result of this implementation.

Next steps:
- [x] ~~now most of the methods that allow writing to the block store always overwrite data. Introduce an option to disable overwrite.~~ After the latest discussion, we decided that overwriting will be implementation specific and is left up to the implementer of `DataWriter<K, T>`
- [x] wire up the implementation in the node storage component as a replacement.
- [x] refactor tests to use the new block store from casper-storage

### CODE REVIEW

File: node/src/components/storage.rs [hunk 1]
```diff
@@ -154,37 +125,9 @@ const STORAGE_FILES: [&str; 5] = [
 pub struct Storage {
     /// Storage location.
     root: PathBuf,
-    /// Environment holding LMDB databases.
+    /// Block store
     #[data_size(skip)]
```
Comment (Fraser999) [inline node/src/components/storage.rs]: Seems like we don't need to skip `data_size` on the `IndexedLmdbBlockStore`?

File: storage/src/block_store/lmdb/temp_map.rs [hunk 2]
```diff
@@ -0,0 +1,73 @@
+use std::collections::BTreeMap;
+
+#[allow(dead_code)]
```
Comment (Fraser999) [inline storage/src/block_store/lmdb/temp_map.rs]: Can this `allow` instance and the two others in this file be removed?

File: storage/src/block_store/lmdb/versioned_databases.rs [hunk 3]
```diff
@@ -267,6 +275,7 @@ where
 
     /// Returns the keys from the `current` database only.
     #[cfg(test)]
+    #[allow(dead_code)]
```
Comment (Fraser999) [inline storage/src/block_store/lmdb/versioned_databases.rs]: Looks like this method can be deleted now?

**Review by Fraser999 — APPROVED** [review]
No blockers, basically LGTM.

File: storage/src/block_store/lmdb/versioned_databases.rs (hunk 3 above)
Comment (alsrdn) [reply storage/src/block_store/lmdb/versioned_databases.rs]: removed in https://github.com/casper-network/casper-node/pull/4498/commits/574e42a50f3b93328b9e8211c025bda934d80225

File: storage/src/block_store/lmdb/temp_map.rs (hunk 2 above)
Comment (alsrdn) [reply storage/src/block_store/lmdb/temp_map.rs]: removed in https://github.com/casper-network/casper-node/pull/4498/commits/574e42a50f3b93328b9e8211c025bda934d80225

File: node/src/components/storage.rs (hunk 1 above)
Comment (alsrdn) [reply node/src/components/storage.rs]: removed skip in https://github.com/casper-network/casper-node/pull/4498/commits/574e42a50f3b93328b9e8211c025bda934d80225

**Review by EdHastingsCasperAssociation — APPROVED** [review]

**Comment by alsrdn** [comment]
bors merge

**Pull request merged**